common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java

```diff
@@ -95,19 +110,140 @@ private ITestRun getTestRun() {
         Assert.notNull(parser, "Found no matched result parser: " + performanceTestResult.parserType);
         return parser.parse(performanceTestResult);
     }
+
+    @Override
+    public void testRunStarted() {
+        inspectWithLifeCycle(InspectionStrategy.WhenType.TEST_RUN_STARTED, "");
+    }
+
+    @Override
+    public void testRunFinished() {
+        inspectWithLifeCycle(InspectionStrategy.WhenType.TEST_RUN_FINISHED, "");
+    }
+
+    @Override
+    public void testStarted(String testName) {
+        inspectWithLifeCycle(InspectionStrategy.WhenType.TEST_STARTED, testName);
+    }
+
+    @Override
+    public void testSuccess(String testName) {
+        inspectWithLifeCycle(InspectionStrategy.WhenType.TEST_SUCCESS, testName);
+    }
+
+    @Override
+    public void testFailure(String testName) {
+        inspectWithLifeCycle(InspectionStrategy.WhenType.TEST_FAILURE, testName);
+    }
+
+    public void testTearDown(DeviceInfo deviceInfo) {
+        ITestRun testRun = getTestRun();
+
+        List<PerformanceTestResult> resultList = parseForTestRun(testRun);
+        savePerfTestResults(resultList);
+
+        List<ScheduledFuture<?>> timerList = inspectPerformanceTimerMap.get(testRun.getId());
+        if (timerList != null) {
+            for (ScheduledFuture<?> timer : timerList) {
+                timer.cancel(true);
+            }
+        }
+
+        inspectPerformanceTimerMap.remove(testRun.getId());
+        testLifeCycleStrategyMap.remove(testRun.getId());
+        testRunPerfResultMap.remove(testRun.getId());
+
+        //TODO Android battery: adb shell dumpsys battery reset using Device info
+    }
+
+    private void inspectWithLifeCycle(InspectionStrategy.WhenType whenType, String description) {
+        List<InspectionStrategy> strategyList = testLifeCycleStrategyMap.get(getTestRun().getId());
+        if (strategyList == null) return;
+
+        for (InspectionStrategy inspectionStrategy : strategyList) {
+            if (inspectionStrategy == null) continue;
+            PerformanceInspection inspection = inspectionStrategy.inspection;
+            if (inspectionStrategy.when == null) continue;
+
+            if (whenType == InspectionStrategy.WhenType.TEST_RUN_STARTED) {
+                // initialize inspector
+                PerformanceInspection initialInspection = new PerformanceInspection(
+                        InspectionStrategy.WhenType.TEST_RUN_STARTED.name(),
+                        inspection.inspectorType, inspection.appId, inspection.deviceIdentifier, true);
+                inspect(initialInspection);
+            } else if (inspectionStrategy.when.contains(whenType)) {
+                PerformanceInspection lifeCycleInspection = new PerformanceInspection(
+                        whenType.name() + "-" + description,
+                        inspection.inspectorType, inspection.appId, inspection.deviceIdentifier, false);
+                inspect(lifeCycleInspection);
+            }
+        }
```

Comment:
Seem the only difference is the name? Could we use the ternary operator to do this?

Reply:
Simplify the code, thx